Handle packages with binaries

rygine opened the pull request:
Minor refactor of some methods for easier to read code using `async.series`

rygine commented:
At my company, we've written a few CLI tools that are accessed via NPM scripts in each package. Bootstrapping did not factor this in and they would fail to run.

thejameskyle commented on src/commands/BootstrapCommand.js:
```diff
@@ -49,24 +50,25 @@ export default class BootstrapCommand extends Command {
 
-      this.createLinkedDependency(linkSrc, linkDest, dependency.name, done);
+      this.createLinkedDependency(linkSrc, linkDest, dependency.name, (err) => {
+        const bin = dependency._package.bin;
```
This should check `err` at the beginning and call `done(err)` if it exists and stop

rygine replied:
fixed

rygine commented:
i'm noticing that the unit tests are non-deterministic, both locally and travis

paulyoung commented:
This is sorely needed! Anything I can do to help move this along?

rygine commented:
@paulyoung, this functionality is actually covered in PR #237 as well.

gigabo commented on src/commands/BootstrapCommand.js:
```diff
@@ -49,24 +50,24 @@ export default class BootstrapCommand extends Command {
-      this.createLinkedDependency(linkSrc, linkDest, dependency.name, done);
+      this.createLinkedDependency(linkSrc, linkDest, dependency.name, (err) => {
+        if (!err && dependency._package.bin) {
+          const srcBinLocation = path.join(linkSrc, dependency._package.bin);
```
Would be nice to add a `bin` getter in src/Package.js for consistency. Then this could be:

```javascript
const srcBinLocation = path.join(linkSrc, dependency.bin);
```

gigabo replied:
Actually, `dependency.bin` [isn't guaranteed to be a string](https://github.com/lerna/lerna/pull/192/files#r68466171).

Could normalize inside the getter to always return an object.  Else the logic here will need some rework.

gigabo commented on src/commands/BootstrapCommand.js:
```diff
@@ -49,24 +50,24 @@ export default class BootstrapCommand extends Command {
 
-      this.createLinkedDependency(linkSrc, linkDest, dependency.name, done);
+      this.createLinkedDependency(linkSrc, linkDest, dependency.name, (err) => {
+        if (!err && dependency._package.bin) {
```
`bin` can be [an object](https://docs.npmjs.com/files/package.json#bin).  Should probably support that here.

rygine replied:
i've neglected this PR in favor of PR #237, which accounts for this and also add a `.bin` getter to `Package` as you mentioned above.

gigabo replied:
Great!  So it should be easy to back port. 😁 

The functionality provided by _this_ PR is independently useful and, frankly, less controversial than #237.  I don't want it to get delayed by association.  I'd like to see this merge, independently, as soon as possible.  That would have the added benefit of reducing the scope of #237, which is pretty big.

rygine replied:
Agreed. Will update.

rygine commented:
@gigabo This should do it. I also uncovered a potential issue with the `ignore` flag during bootstrap. Fixed that as well.

* Added `symlink` method to FileSystemUtilities
* Added `bin` property to Package
* Updated bootstrap to factor in dependencies with multiple binaries
* Added test for package with multiple binaries
* Refactored way ignored packages are determined (to be more like run command)
* Updated unit test for ignored package

When a package is ignored while bootstrapping, it should be completely taken out of the equation. This update alters the previous behavior of not bootstrapping an ignored package, but still treating it as though it was part of the process. If a package is ignored, it should be treated as an external dependency. The updated unit test reflects this.

gigabo commented:
> If a package is ignored, it should be treated as an external dependency.

Hmm... that's actually not what I'd naively expect.  😕 

I'd expect "ignore package foo" to skip doing anything in `packages/foo`, but to do the _normal_ things in `package/bar`, including linking `foo` as an in-repo dep.  Am I just misunderstanding the `--ignore` flag?

rygine commented:
@gigabo that's a good question. what exactly does ignore mean?

my initial thought was that it doesn't get any of its dependencies "linked" in the bootstrap, but i'm not sure that makes sense.

for example, check out this scenario:

* packageA depends on packageB
* packageC depends on packageA

i run bootstrap and ignore packageA, which means that packageA does not have a "linked" or installed version of packageB. meanwhile, i bootstrap packageC and "link" to packageA like normal. now, when i import packageA into packageC, packageB will not be imported as no form of it exists in `packageA/node_modules/`.

under the new scenario, packageA would be installed as an external dependency of packageC and all would be well. i don't see much use for the `ignore` flag during bootstrap except for some edge-cases when you simply want to disregard a module.

it is a bit confusing to think about. once i realized how things would be bootstrapped when you ignore a package, i'm not sure the original way of doing it would actually work.

gigabo commented:
@rygine - The behavior or `--ignore` is worth examining, but seems unrelated to handling packages with binaries.  Can that part be pulled out into a separate issue or PR?

rygine commented:
@gigabo makes sense. amended the commit.

rygine commented on src/commands/BootstrapCommand.js:
```diff
@@ -28,16 +42,11 @@ export default class BootstrapCommand extends Command {
-    const todoPackages = PackageUtilities.filterPackages(this.packages, ignore, true);
-
-    // Get a trimmed down graph that includes only those packages.
-    const filteredGraph = PackageUtilities.getPackageGraph(todoPackages);
```
this is slightly unrelated to the PR, but i moved the package filtering to the `initialize` method to mimic the behavior of `lerna run` for consistency.

gigabo replied:
> unrelated to the PR

Generally speaking, smaller more targeted PRs are easier to review and therefore tend to get merged more quickly.

This PR adds much needed functionality, but it has gone through a number of iterations that each included a different set of unrelated changes.  These other changes may _also_ be valuable, but it would be easier to consider them on their own merit if they were showcased in dedicated PRs.

rygine replied:
point taken. amended.

gigabo replied:
Thanks, that makes it easier to read. 👍

rygine commented on src/commands/BootstrapCommand.js:
```diff
@@ -28,16 +42,11 @@ export default class BootstrapCommand extends Command {
 
     // As packages are completed their names will go into this object.
-    const donePackages = {};
+    this.donePackages = {};
```
want to keep track of bootstrapped packages within the class to display the number at the end, see L#35.

gigabo commented on src/commands/BootstrapCommand.js:
```diff
@@ -55,12 +55,28 @@ export default class BootstrapCommand extends Command {
+        // if package has dependencies with binaries
+        if (depsWithBin.length) {
+          // symlink binaries
+          actions.push(cb => this.symlinkDependencyBinaries(pkg, depsWithBin, cb));
```
I think the binaries should be linked _before_ `runPrepublishForPackage`.

A package may have a `prepublish` script that uses a binary from a dependency.

rygine replied:
make sense.

gigabo commented on src/commands/BootstrapCommand.js:
```diff
@@ -128,13 +136,49 @@ export default class BootstrapCommand extends Command {
+    const actions = [];
+    dependencies.forEach(dependency => {
+      const { location, name, bin } = dependency;
+      const destPath = path.join(pkg.nodeModulesLocation);
```
What does this `path.join` do?

rygine replied:
good catch. this was leftover from some previous code.

gigabo commented on src/FileSystemUtilities.js:
```diff
@@ -51,6 +51,17 @@ export default class FileSystemUtilities {
   }
 
+  @logger.logifyAsync
+  static symlink(src, dest, callback, type = "file") {
```
It doesn't look like the caller uses the `type` parameter?

All other asynchronous methods in `FileSystemUtilities` take the callback as the _final_ argument.

rygine replied:
ahh yes. will fix.

rygine commented:
@gigabo I addressed your comments with an amended commit. For future reference, should I continue to amend commits, or just make another one for things like this?

gigabo commented:
> should I continue to amend commits, or just make another one for things like this?

From my perspective it's nice to have a new commit so I can see clearly what changed.  PRs are typically squashed during merge, so it won't clutter the commit history on master.

gigabo commented:
@hzoo, @thejameskyle - Anything else you'd like to see here?  I want to use this!

gigabo commented:
Taking https://github.com/lerna/lerna/pull/300#issuecomment-238647550 as a 👍  here.

Thanks for sticking with this @rygine.  This is a big improvement!
